plugins/telemetry-otel/src/main/java/org/opensearch/telemetry/tracing/processor/OTelSpanProcessor.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.telemetry.tracing.processor;
+
+import org.opensearch.telemetry.tracing.attributes.SamplingAttributes;
+
+import java.util.Objects;
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.sdk.common.CompletableResultCode;
+import io.opentelemetry.sdk.trace.ReadWriteSpan;
+import io.opentelemetry.sdk.trace.ReadableSpan;
+import io.opentelemetry.sdk.trace.SpanProcessor;
+
+/**
+ * Implementation of the SpanProcessor and delegates to the configured processor.
+ */
+public class OTelSpanProcessor implements SpanProcessor {
+
+    private final SpanProcessor delegateProcessor;
+
+    /**
+     * *
+     * @param delegateProcessor the span processor to which this processor will delegate
+     */
+    public OTelSpanProcessor(SpanProcessor delegateProcessor) {
+        this.delegateProcessor = delegateProcessor;
+    }
+
+    /**
+     * Called when a {@link Span} is started, if the {@link
+     * Span#isRecording()} returns true.
+     *
+     * <p>This method is called synchronously on the execution thread, should not throw or block the
+     * execution thread.
+     *
+     * @param parentContext the parent {@code Context} of the span that just started.
+     * @param span          the {@code Span} that just started.
+     */
+    @Override
+    public void onStart(Context parentContext, ReadWriteSpan span) {
+        this.delegateProcessor.onStart(parentContext, span);
+    }
+
+    /**
+     * Returns {@code true} if this {@link SpanProcessor} requires start events.
+     *
+     * @return {@code true} if this {@link SpanProcessor} requires start events.
+     */
+    @Override
+    public boolean isStartRequired() {
+        return this.delegateProcessor.isStartRequired();
+    }
+
+    /**
+     * Called when a {@link Span} is ended, if the {@link
+     * Span#isRecording()} returns true.
+     *
+     * <p>This method is called synchronously on the execution thread, should not throw or block the
+     * execution thread.
+     *
+     * @param span the {@code Span} that just ended.
+     */
+    @Override
+    public void onEnd(ReadableSpan span) {
+        if (span != null
+            && span.getSpanContext().isSampled()
+            && Objects.equals(
+                span.getAttribute(AttributeKey.stringKey(SamplingAttributes.SAMPLER.getValue())),
+                SamplingAttributes.INFERRED_SAMPLER.getValue()
+            )) {
+            if (span.getAttribute(AttributeKey.booleanKey(SamplingAttributes.SAMPLED.getValue())) != null) {
+                this.delegateProcessor.onEnd(span);
```

Comment:
How come we are not calling `delegateProcessor.onEnd(span)` in all branches? We don't know what `delegateProcessor` is actually doing, it may track spans with onStart  / onEnd call

Reply:
Delegated Processor here is the BatchSpanProcessor itself. This custom processor is acting as an intermediate layer between BatchSpanProcessor and Otel framework and forwards the normal span i.e. spans which are not marked by inferred sampler directly to BatchSpanProcessor and if the span is marked by inferred sampler then filter out spans that are outlier or their children are outlier. This is done by adding a sampled attribute to the span itself so that the processor can distinguish between the outlier and non-outlier spans

Reply:
You are correct @reta . However we cannot use BatchSpanProcessor as it-is because we want the processor to be attribute aware. What we can do is create a WrappedBatchSpanProcessor that will always initialize BatchSpanProcessor as the delegated processor and use the same.

Reply:
Since we are not following the specs (red flag here), the only way implement this feature is to provide our own `SpanProcessor` that does not rely on any existing OTEL provided classes.

Reply:
So as I understand the concern here is that the onStart for certain spans will be called however the onEnd will not be called which could be an issue . Is that correct? So it would be better to have our own implementation for the same?

Reply:
Yes, it violates specs and leaks spans